File: Java/ds/MiddleOfLinkedList.java
//Java program to find middle of LinkedList

/*Possible Approaches: 
i) Traverse List: Traverse list to find it's length, then scan list to locate n/2 node from the beginning.
ii) Two - Pointer Approach: It is the efficient approach (only one scan). Use 2 pointers. Move one at twice the speed of second. 
When first pointer reaches the end of list, second pointer will point the middle node of the LinkedList.					   		
*/
import java.util.*;
public class MiddleOfLinkedList {
	Node head;
    	static class Node {
		int data;
		Node next;
			Node(int data) {
            			this.data = data;
			}
    		}
 
    public void addToTheLast(Node node) {
 
        if (head == null) {
            head = node;
        }
	else {
            Node temp = head;
            while (temp.next != null)
                temp = temp.next;
		temp.next = node;
        }
    }
 
    public void printList() {
        Node temp = head;
        while (temp != null) {
            System.out.format("%d ", temp.data);
            temp = temp.next;
        }
        System.out.println();
    }
 
    // This function will find middle element in linkedlist
    public Node findMiddleNode(Node head){
    	Node pointer1, pointer2;       
    	pointer1 = pointer2 = head;	 
    	int i = 0;
    	while(pointer1.next != null){
    		if(i == 0){
    			pointer1 = pointer1.next;
    			i=1;
    		}
    		else if(i == 1){
    			pointer1 = pointer1.next  
    			pointer2 = pointer2.next;
    			i=0;
    		}
    	}	
    	return pointer2; 
    }
    	
    //Main Method
    public static void main(String[] args) {
        MiddleOfLinkedList list = new MiddleOfLinkedList();
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter the size of Linked List:");
        int n = sc.nextInt();
        System.out.println("Enter the data of Linked List:");
        int num = sc.nextInt();
        Node head = new Node(num);
        list.addToTheLast(head);
        Node tail = head;
        for(int i=0; i<n-1; i++){
            num = sc.nextInt();
            tail.next = new Node(num);
            tail = tail.next;
        }
        list.printList();
        Node middle = list.findMiddleNode(head);
        System.out.println("Middle node is: "+ middle.data);
    }
}

/*OUTPUT
  Enter the size of Linked List:
  5
  Enter the data of Linked List:
  1 2 3 4 5
  1 2 3 4 5 
  Middle node is: 3
  
  Time Complexity: O(n) 
  Space Complexity: O(1)
 */

File: C-Plus-Plus/ds/Middle_of_Linked_list.cpp
/*

Given N, Linked List 
Your task is to print the middle element(Data) of Given Linked List 
Example:
N=5
1 2 3 4 5
Answer should be 3

*/
#include <bits/stdc++.h>
using namespace std;

class Node
{
public:
    int data;
    Node *next;
};
void push(Node **head, int data)
{
    Node *new_node = new Node();
    new_node->data = data;
    (new_node)->next = (*head);
    (*head) = new_node;
}
void printLL(Node *temp)
{
    while (temp != NULL)
    {
        cout << temp->data << "->";
        temp = temp->next;
    }
}
int middleinLL(Node *head)
{
    Node *slow = head, *fast = head;
    while (fast && fast->next)
    {
        slow = slow->next;
        fast = fast->next->next;
    }
    return slow->data;
}
int main()
{
    Node *head = NULL;
    int n;
    cout << "Enter Numbers of Element N:";
    cin >> n;
    cout << "Enter elements:";
    for (int i = 0; i < n; i++)
    {
        int a;
        cin >> a;
        push(&head, a);
    }
    cout << "Linked List:";
    printLL(head);
    cout << "\nMiddle Element of the Linked List:" << middleinLL(head);
}

/*

Sample Input/Output:

Input:
Enter Numbers of Element N:7
Enter elements:4 1 6 2 4 9 1

Output:
Linked List:1->9->4->2->6->1->4->
Middle Element of the Linked List:2

Time-Complexity: O(n)
Space-Complexity: O(1)

*/
